### DISCUSSION TITLE
Entwicklungssystem

### DISCUSSION THREAD

**Discussion by Weseler485** [OpenCCU/OpenCCU#2231] [Development]
naHallo Leute,
ich möchte mich etwas mehr mit der  Entwicklung beschäftigen und habe dazu zwei grundlegende Fragen:

1. Wie habt ihr eure Entwicklungs-/Testumgebung realisiert? Ein leeres System oder eine Kopie des Produktiven? Aber wie erreicht ihr dann, dass dieses nicht Kontakt mit den Geräten aufnimmt?

2. Die Verzeichnisse mit den Patches und Co sind ja in dem normalen Images nicht drin. Startet ihr die Programme zum Generieren der diffs nicht auf der CCU sondern in einem anderen Linux oder werden die ganzen Ordner immer zuvor in die Entwicklungsumgebung kopiert?

Denke das sollte erstmal das wichtigste gewesen sein...

Danke euch!!
Grüße 
Maik2208

**Comment by jens-maus** [comment]
> Hallo Leute, ich möchte mich etwas mehr mit der Entwicklung beschäftigen und habe dazu zwei grundlegende Fragen:

Gut zu lesen! :)

> 1. Wie habt ihr eure Entwicklungs-/Testumgebung realisiert? Ein leeres System oder eine Kopie des Produktiven? Aber wie erreicht ihr dann, dass dieses nicht Kontakt mit den Geräten aufnimmt?

Ich entwickle immer auf einem OVA-basierten Testsystem das ich unter Proxmox betreibe. Dafür einen Pi/Tinkerboard zu nutzen ist schon aus geschwindigkeitsgründen keine gute Idee. So eine OVA-basiertes RaspberryMatic ist einfach superschnell eingerichtet und dann auch schneller geupdatet, usw. Und dann hab ich eben einen gewissen Testpark an HomeMatic/homematicIP Geräten der natürlich nicht in meiner produktiven Umgebung genutzt wird und damit teste ich dann was ich so kann und entwickle darauf basierend.

> 2. Die Verzeichnisse mit den Patches und Co sind ja in dem normalen Images nicht drin. Startet ihr die Programme zum Generieren der diffs nicht auf der CCU sondern in einem anderen Linux oder werden die ganzen Ordner immer zuvor in die Entwicklungsumgebung kopiert?

Mit "Patches" meinst du sicherlich die WebUI patches. Die haben natürlich nichts auf der CCU/RaspberryMatic Zentrale selbst zu tun. Die Entwicklung findet hier rein unter einem lokalen Ubuntu/Linux (amd64) statt und der Ablauf ist dann wie folgt:

1. Wenn ich Anpassungen an der WebUI mache setze ich direkt in den jeweiligen Dateien unter der Test-CCU/RaspberryMatic um und lege vorher die entsprechenden *.orig Dateien an um dann später einfacher ein diff machen zu können.
2. Wenn die Änderungen funktionieren und ich sicher bin das alles umgesetzt wurde kopiere ich die geänderten Dateien in mein Ubuntu/Linux Entwicklungssystem rüber wo das RaspberryMatic GitHub ausgecheckt liegt und dann pflege ich das in einen WebUI Patch entsprechend ein und lasse dann mit `create_patches.sh` die entsprechende `.patch` Dateien generieren.
3. Dann lass ich das gesamte RaspberryMatic nochmal neu erstellen und wenn das entsprechende update zip fertig gebaut ist spiele ich das in meine Test-CCU/RaspberryMatic nochmal via WebUI Update Mechansimus ein und kontrolliere dann das die von mir vorgenommenen Änderungen wirklich im Build-System korrekt angekommen sind.

Hoffe das klärt die ersten Fragen die du hattest.

**Comment by Baxxy13** [comment]
Kommt ja auch immer drauf an was man machen will. 
Reine Optimierungen / Fixes / Erweiterungen die nix mit Aktoren / Sensoren zu tun haben macht man am besten auf einem jungfräulichen Testsystem. Ich nutze primär eine VM, ab und zu auch nen Pi4B mit SSD.

Ist es was Gerätespezifisches dann sollte auch das entsprechende Gerät im Testsystem angelernt sein.

Die Umsetzung läuft dann hier auch etwa so wie Jens in Punkt 1 & 2 schrieb.
Ich nutze aber ein Debian in einer Virtualbox-VM unter Windows. 
Da ich nicht so der Kommandozeilen-Typ bin ist im Debian der Github-Desktop Client installiert.

**Comment by Weseler485** [comment]
Danke euch beiden! 

Grundlegend lese ich erstmal zwei Dinge raus:
1. Entwicklungsumgebung ist grüne Wiese, maximal ein paar dedizierte Testgeräte angelernt, aber nicht wie ich es auch anderen Ecken gewohnt bin eine Kopie des Prodsystems.

2. Es ist doch so wie vermutet, dass ich mit einem weiteren Linuxsystem arbeiten muss um die Scripte auszuführen. Hatte gehofft, dass ich mir das ersparen kann. Muss ich mir mal Gedanken machen, wie ich das am besten umsetze.

Prinzipiell geht es mir erstmal um Kleinigkeiten umzusetzen um mich mehr mit der Materie zu beschäftigen. Wie zB das Ticket mit dem doppelten "bei" in den Bedingungstexten. Aber auch Erweiterungen, die mir persönlich den Umgang schöner machen, hier im Projekt aber nicht gewünscht waren weil zu großer Eingriff in wie webui.js

**Comment by jp112sdl** [comment]
Bei mir sind es 2 Systeme.
Ein Pi3 mit HM-MOD-RPI-PCB, der irgendwo in der Ecke liegt und eine OVA.

Daheim mit Notebook-Akkubetrieb arbeite ich auf dem Pi3. 

Wenn ich am Netzteil hänge, dann mit der OVA (Virtualbox erzeugt ja doch ein bisschen Last) und angestecktem HB-RF-USB oder HB-RF-HAP.

Das Git-Repo liegt direkt auf meinem Mac und auch das Patch-Erstellen mache ich direkt auf diesem, ohne zusätzliches Linux (Ubuntu o.ä.). Da ist der Vorteil, dass macOS auf Unix basiert.

Für letzteres (_> Aber auch Erweiterungen, die mir persönlich den Umgang schöner machen_) habe ich mir ein kleines [Addon](https://github.com/jp112sdl/CCUstomize) gebaut, dass ein paar Patch nachfährt.

**Reply by mbhomie007** [reply]
Nebenbei: 
"Hinzufügen eines Menüpunkts "Systemsteuerung" mit den fixen Elementen aus "Einstellungen->Systemsteuerung"

Die Erweiterung finde ich klasse, so hat man immer ein Direktlink. 👌
Und optisch auch sehr ansprechend.

**Reply by jens-maus** [reply]
Besser wäre noch jemand mit genug Motivation und Web-Design KnowHow setzt sich jetzt mal hin und fängt an alles auf Bootstrap JS umzustellen.. dann fängt es erst richtig an Spass zu machen :)

**Comment by Weseler485** [comment]
Spricht eigentlich etwas dagegen, ein Backup des Produktivsystems dort einzuspielen?
Also kommuniziert das System mit einem EQ3-Server und legt dann meine produktive Instanz lahm?

PS:
Habt ihr paar nützliche Befehle parat, die man beim Entwickeln nutzen kann?
Wie kann ich z.B. den Webserver neustarten wenn ich Änderungen vorgenommen habe und diese in der GUI nicht aktiv werden?

**Reply by jens-maus** [reply]
> Spricht eigentlich etwas dagegen, ein Backup des Produktivsystems dort einzuspielen? Also kommuniziert das System mit einem EQ3-Server und legt dann meine produktive Instanz lahm?

Wenn du homematicIP Geräte hast, dann definitiv. Sobald du dein Backup irgendwo anders einspielst wo ein anderes Funkmodul zur verfügung steht wird ja das rekeying passieren und dann kann dein produktivsystem nicht mehr auf die geräte zugreifen. Und bei BidCos-RF ist es ähnlich, weil dann funkt das zweite System ja auf der gleichen Funkadresse mit den Geräten und dann kommt alles durcheinander.

Deshalb für ein Entwicklungsystem immer ein separates System/Installation und auch separate Testgeräte nutzen wenn du keinerlei negativen Effekt auf dein Produktivsystem haben willst. Es kann ja auch z.B. sein das du da irgendwelche Skripte laufen hast die im selben Netzwerk dann mit anderen Dingen deines Haushaltes (z.B. ioBroker, Philips Hue, etc.) sprechen will und dann beeinflusst das Entwicklungs/Testsystem ja auch diese parallel zum Produktivsystem.

> PS: Habt ihr paar nützliche Befehle parat, die man beim Entwickeln nutzen kann? Wie kann ich z.B. den Webserver neustarten wenn ich Änderungen vorgenommen habe und diese in der GUI nicht aktiv werden?

Webserver neustarten: `monit restart lighttpd`
ReGaHss neustarten: `monit restart regahss`

Wenn du änderungen an der WebUI vornimmst kommt es allerdings drauf an **was** du modifiziert hast. Ist es nur ein `*.js` dann sollte das nach browser cache löschen sofort ohne neustart angewendet werden. Für anderen Dateien wie `*.htm` Dateien oder `*.cgi` skripte kann es allerdings sein das du zumindest ReGaHss neustarten musst damit dieses die Dateien frisch einliest und nicht deren Cache verwendet. Aber auch das ist nicht immer so: Ist eben "try and error" und "probieren geht über studieren" :)

**Reply by Weseler485** [reply]
Ich scheitere seit Stunden an einer Kleinigkeit, für dich meine Kenntnisse leider nicht ausreichen :(

```
<%
string s = "${Textfeld}";
%>
jQuery("#Platzhalter").html("<% Write(s); %>");
```

Es wird "${Textfeld}" ausgegeben. Wie bekomme ich das Textfeld korrekt übersetzt? 

```
<%
string s = translateKey("Textfeld");
%>
jQuery("#Platzhalter").html("<% Write(s); %>");
```
Funktioniert nicht, da das JS dann hängen bleibt.
